### ISSUE TITLE
[Request] React-like ref for widgets

### ISSUE THREAD

**Issue by rrousselGit** [flutter/flutter#35742]
## Use case

`GlobalKey` allows other widgets to access to the `State` of a widget. But it comes with many issues:

- It's possible to read the content of a `GlobalKey` inside `build`, which is a very bad practice
- the documentation states that it is not performant
- It's difficult to make `GlobalKey` compile-time safe. There's nothing preventing the following:
  ```dart
  Navigator(key: GlobalKey<FormState>()) // <FormState> doesn't make sense
  ```
- it can't be used inside a `StatelessWidget` easily without losing composition, example:
  ```dart
  class Foo extends StatelessWidget {
    final formKey = GlobalKey<FormState>();
  
    @override
    Widget build(BuildContext context) {
      return Form(
        key: formKey,
        child: ...,
      );
    }
  }
  ```
  Such code is bad because if `Foo` gets recreated, then the `GlobalKey` will be recreated, which will destroy the form.

## Proposal

This proposal is about implementing an alternative to `GlobalKey`, to access the state of descendants, that is **not** a `Key`.

It'd be expressed through two things:

- A class `Ref` like so:

  ```dart
  class Ref<T extends State> {
    T get state;
  }
  ```

- A new property on `StatefulWidget` is added:

  ```dart
  abstract class StatefulWidget {
    const Widget({Key key, Ref ref});
  }
  ```

then all `StatefulWidget` with a public `State` are expected to expose an optional `ref`.

For example, `Form` would do:

```dart
class Form extends StatefulWidget {
  Form({Key key, Ref<FormState> ref}): super(key: key, ref: ref);
}
```

Which means a `StatelessWidget` can then do: (without `FormState` being destroyed if the ref change)

```dart
class Foo extends StatelessWidget {
  final formRef = Ref<FormState>();
  
  @override
  Widget build(BuildContext context) {
    return Form(
      ref: formRef,
      onChanged: () {
        // do something with formRef.state
      }
      child: ...,
    );
  }
}
```

It'd come with a few rules:

- Accessing `Ref.state` inside `build`, should be strictly forbidden. This can be done through `BuildOwner.debugBuilding`
- The `Ref` associated with a `StatefulWidget` can freely be replaced by another one without having any impact on `State`, as opposed to `Key`
- Subclasses of `StatefulWidget` are expected to specify `T` of `Ref<T>`:
  DON'T
  ```dart
   Foo({Ref ref}): super(ref: ref);
  ```
  DO:
  ```dart
  Foo({Ref<FooState> ref}): super(ref: ref);
  ```

______

This solves all the previously mentioned points because:

- it is a lot more performant then `GlobalKey` since it does nothing else besides storing a reference to `State`.
- it voluntarily can't be read inside `build`
- it can be used inside `StatelessWidget`
- it can be made type-safe, such that the following won't compile:
  ```dart
   Form(ref: Ref<NavigatorState>())
  ```

**Comment by slightfoot** [comment]
I agree with and have seen all the bad use-cases. However I am not sure adding another parameter and type would resolve the issue and not make things more complicated. Especially when it does a very similar job to GlobalKey. I think it would better to fix the issues you explained in the current framework. This then might make it possible to have a RefKey, or something that is similar to what is suggested without the need for even more parameters.

**Comment by rrousselGit** [comment]
> This then might make it possible to have a RefKey

No, because the root of some of these issues is because we're using keys for something they're not meant to.

As long as we take the key slot to obtain the state of descendants, then both the type-safety and "GlobalKey inside StatelessWidget" will stay.
